drivers/Aqara/aqara-presence-sensor/src/lunchbox/rest.lua

```diff
@@ -0,0 +1,378 @@
+---@field package _received_body boolean
+---@field package _parsed_headers boolean
+---@field public new fun(status_code: number, socket: table?): ChunkedResponse
+---@field public fill_body fun(self: ChunkedResponse): string?
+---@field public append_body fun(self: ChunkedResponse, next_chunk_body: string): ChunkedResponse
+
+local socket = require "cosock.socket"
+local utils = require "utils"
+local lb_utils = require "lunchbox.util"
+local Request = require "luncheon.request"
+local Response = require "luncheon.response" --[[@as ChunkedResponse]]
+local api_version = require("version").api
+
+local RestCallStates = {
+  SEND = "Send",
+  RECEIVE = "Receive",
+  RETRY = "Retry",
+  RECONNECT = "Reconnect",
+  COMPLETE = "Complete",
+}
+
+local function connect(client)
+  local port = 80
+  local use_ssl = false
+
+  if client.base_url.scheme == "https" then
+    port = 443
+    use_ssl = true
+  end
+
+  local sock, err = client.socket_builder(client.base_url.host, port, use_ssl)
+
+  if sock == nil then
+    client.socket = nil
+    return false, err
+  end
+
+  client.socket = sock
+  return true
+end
+
+local function reconnect(client)
+  if client.socket ~= nil then
+    client.socket:close()
+    client.socket = nil
+  end
+  return connect(client)
+end
+
+---comment
+---@param client RestClient
+---@param request HttpMessage
+---@return integer? bytes_sent
+---@return string? err_msg
+---@return integer idx
+local function send_request(client, request)
+  if client.socket == nil then
+    return nil, "no socket available", 0
+  end
+  local payload = request:serialize()
+
+  local bytes, err, idx = nil, nil, 0
+
+  repeat bytes, err, idx = client.socket:send(payload, idx + 1, #payload) until (bytes == #payload)
+    or (err ~= nil)
+
+  return bytes, err, idx
+end
+
+---@param original_response Response
+---@param sock table
+---@return Response?
+---@return string? err
+local function parse_chunked_response(original_response, sock)
+  local ChunkedTransferStates = {
+    EXPECTING_CHUNK_LENGTH = "ExpectingChunkLength",
+    EXPECTING_BODY_CHUNK = "ExpectingBodyChunk",
+  }
+
+  local full_response = Response.new(original_response.status, nil) --[[@as ChunkedResponse]]
+
+  for header in original_response.headers:iter() do full_response.headers:append_chunk(header) end
+
+  local original_body, err = original_response:get_body()
+  if original_body == nil or err ~= nil then
+    return nil, (err or "unexpected nil in error position")
+  end
+  local next_chunk_bytes = tonumber(original_body, 16)
+  local next_chunk_body = ""
+  local bytes_read = 0;
+
+  local state = ChunkedTransferStates.EXPECTING_BODY_CHUNK
+
+  repeat
+    local pat = nil
+    local next_recv, next_err, partial = nil, nil, nil
+
+    if state == ChunkedTransferStates.EXPECTING_BODY_CHUNK then
+      pat = next_chunk_bytes
+    else
+      pat = "*l"
+    end
+
+    next_recv, next_err, partial = sock:receive(pat)
+
+    if next_err ~= nil then
```

Comment:
This should handle wantread.

```
if next_err ~= nil then
    if string.lower(next_err) == "closed" then
        if partial ~= nil and #partial >= 1 then
          full_response:append_body(partial)
          next_chunk_bytes = 0
        else
          return nil, next_err
        end
      elseif string.lower(next_err) == "wantread" then
        -- Just wait and try again if socket is not ready yet
         socket.sleep(0.1)  -- Short sleep to prevent busy-waiting
      else
        return nil, ("unexpected error reading chunked transfer: " .. next_err)
      end
    end
```